Redefine `Coin` in terms of `Natural` and remove the `Bounded` instance.

jonathanknowles opened the pull request:
## Issue Number

ADP-1283

## Summary

This PR:
- Redefines `Coin` in terms of `Natural` (rather than `Word64`).
- Removes the `Bounded` instance for `Coin`.
- Adds `txOut{Min,Max}Coin` constants to `Primitive.Types.Tx`, in the same style as the existing `txOut{Min,Max}TokenQuantity` constants.
- Pushes validation checks for coins that must be within a certain range to the places where those checks are actually required.

## Implementation Notes

Where possible, this PR uses `intCast` and `intCastMaybe` (instead of `fromIntegral`) to perform statically-checked integral conversations.

## Motivation

The wallet uses the `Coin` type to represent an amount of _lovelace_.

It's currently defined in terms of `Word64`, and has a `Bounded` instance that limits the range to `[0, 45_000_000_000_000_000]`.

However, this approach is problematic, for several reasons:

- Boundary checks for `Coin` values make sense in contexts where there are boundaries.  However, for pure `Coin` values in the absence of any particular context, there isn't one particular upper bound that makes sense.
- The current choice of `Bounded` specifically defines the limits of what can be included in a _transaction output_. However, encoding values that appear in transaction outputs is not the only use of the `Coin` type.
- For example, it's quite reasonable to use `Coin` for other purposes, like finding the total volume of ada transacted over some period of time. For such usages, there is no obvious choice of upper limit that we could consider “ideal”.
- Right now, we often use awkward workarounds such as casting `Coin` values to `Natural` values before computing summations, thus losing the descriptive value of using the `Coin` type to mark that “this is a quantity of lovelace”.
- Given that the default data constructor for `Coin` is exported, it’s possible to directly construct “invalid” Coin values from outside the `Coin` module.
- Using the existing `Semigroup` instance, we can combine valid `Coin` values into values that are **_invalid_**, which can lead to unexpected run-time errors when those “invalid” values are passed to functions that we believe to be safe.  💣 

For an example of the last problem, consider this:

```haskell
> :set -XNumericUnderscores
> :set -XOverloadedLists
> import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
> coinA = Coin 45_000_000_000_000_000
> coinB = coinA <> coinA
> coinB
Coin 90000000000000000
> Coin.equipartition coinB [()]
*** Exception: unsafeNaturalToCoin: overflow
CallStack (from HasCallStack):
  error, called at /home/jsk/projects/input-output-hk/cardano-wallet-master/lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs:131:37 in main:Cardano.Wallet.Primitive.Types.Coin
```

jonathanknowles committed:
Add constants `txOutMinCoin` and `txOutMaxCoin`.

The maximum size of a `Coin` value is context dependent.

This commit adds constants that define the limits of a `Coin` value in
the context of a transaction output.

jonathanknowles committed:
Redefine `Coin` in terms of `Natural`.

jonathanknowles committed:
Replace unsafe function `coinQuantity`.

This function is unsafe because `Coin` is now defined in terms of `Natural`.

We replace this function with a pair of functions:

- Coin.toQuantity        :: Integral i => Coin -> Maybe (Quantity n i)
- Coin.unsafeToQuantity  :: Integral i => Coin ->        Quantity n i

jonathanknowles committed:
Replace function `coinToInteger` with safer equivalent.

We use `intCast` instead of `fromIntegral`, which statically checks
that the conversion is safe.

We also use the same naming convention as the existing safe conversion
functions.

jonathanknowles committed:
Replace compatibility function `coinToNatural`.

jonathanknowles committed:
Replace compatibility function `unsafeNaturalToCoin`.

This conversion is no longer unsafe. So we create a safe conversion
function and place it within the "safe conversions" section.

jonathanknowles committed:
Replace outdated comment within `Semigroup` instance for `Coin`.

jonathanknowles committed:
Simplify `FromText` instance for `Coin`.

We can now just rely on the `FromText` instance for `Natural`.

jonathanknowles committed:
Replace `Coin.isValid` with `Tx.coinIsValidForTxOut`.

Boundary checks for `Coin` values make sense in contexts where there
are boundaries.

However, for pure `Coin` values in the absence of any particular
context, there isn't one particular upper bound that makes sense.

In the particular case of a transaction output, `Coin` values must be
within the inclusive range `[Coin 1, Coin 45_000_000_000_000_000]`.

jonathanknowles committed:
Replace `sumCoins` with `Data.Foldable.fold`.

The `Coin` type has both `Semigroup` and `Monoid` instances.

Therefore we can just use `Data.Foldable.fold` to sum coins.

jonathanknowles committed:
Replace `genFixedSize{Coin,TokenBundle}` with `genTxOut{Coin,TokenBundle}`.

These generators are specifically designed to generate values across the
full range of what's permitted to appear in a transaction output.

Therefore, we rename them to have the `genTxOut` prefix, and we move
them to the `Tx.Gen` module.

jonathanknowles committed:
Remove `Bounded` instance for `Coin`.

Instead, we use `txOutMinCoin` and `txOutMaxCoin` where appropriate.

These usages are similar to `txOut{Min,Max}TokenQuantity`.

jonathanknowles committed:
Replace `Coin.addCoin` with `Coin.add`.

Since `Coin` is typically imported in a qualified fashion, we can write
`Coin.add` instead of `Coin.addCoin`.

This change also redefines `<>` for `Coin` in terms of `Coin.add`.

jonathanknowles committed:
Replace `Coin.subtractCoin` with `Coin.subtract`.

Since `Coin` is typically imported in a qualified fashion, we can write
`Coin.subtract` instead of `Coin.subtractCoin`.

Commit:
Regenerate nix

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:
```diff
@@ -297,8 +297,10 @@ log10 = Log10
-computeUtxoStatistics btype =
-    computeStatistics (pure . unCoin . txOutCoin) btype . Map.elems . unUTxO
+computeUtxoStatistics btype
+    = computeStatistics (pure . fromIntegral . unCoin . txOutCoin) btype
```
```suggestion
    = computeStatistics (pure . Coin.unsafeToWord64 . txOutCoin) btype
```

jonathanknowles replied:
Ah, well spotted! I'll fix this.

jonathanknowles replied:
Fixed in 0b95b179480e2d7766eebf369b4616805ef0746d.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3241,8 +3242,8 @@ mkApiFee :: Maybe Coin -> [Coin] -> FeeEstimation -> ApiFee
-    , minimumCoins = coinQuantity <$> minCoins
-    , deposit = coinQuantity $ fromMaybe (Coin 0) mDeposit
+    , minimumCoins = Coin.unsafeToQuantity <$> minCoins
+    , deposit = Coin.unsafeToQuantity $ fromMaybe (Coin 0) mDeposit
```
We're converting to a `Natural`, so this ends up being safe. Can't we use `Coin.toNatural` instead?

Same for minimumCoins.

jonathanknowles replied:
Good point!

I've fixed this in 1937bf7212ef86e67dc7957c5105a62aad5107ef.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs:
```diff
@@ -302,9 +298,7 @@ equipartition
 equipartition c =
-    -- Note: the natural-to-coin conversion is safe, as partitioning guarantees
-    -- to produce values that are less than or equal to the original value.
-    fmap unsafeNaturalToCoin . equipartitionNatural (toNatural c)
```
👍

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs:
```diff
@@ -108,13 +106,7 @@ instance ToText Coin where
-                return x
-            | otherwise =
-                Left $ TextDecodingError "Coin value is out of bounds"
+    fromText = fmap Coin . fromText @Natural
```
This seems like a _slight_ change of behaviour in the CLI https://github.com/input-output-hk/cardano-wallet/blob/a1838496ba697208daf2889ffcd420bea4d1b0a5/lib/cli/src/Cardano/CLI.hs#L1408-L1409

but I guess it makes sense — `stake` is not related to a single TxOut ada value.

jonathanknowles replied:
Yes. I agree this should be acceptable. As you mention, this is unrelated to a transaction output.

If `stake` needs to be constrained further, then we could perhaps make a wrapper type that constrains the value further. Thoughts?

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -953,6 +956,15 @@ mockSealedTx :: HasCallStack => ByteString -> SealedTx
+coinIsValidForTxOut :: Coin -> Bool
+coinIsValidForTxOut c = (&&)
+    (c >= txOutMinCoin)
+    (c <= txOutMaxCoin)
```
Are we sure we couldn't have e.g. a 46B ada TxOut on a custom testnet?

I find this hard-coded value in the ledger specs
https://github.com/input-output-hk/cardano-ledger/blob/4347e54b2827090e898702ccfb93cc731b4a1909/eras/byron/ledger/impl/src/Cardano/Chain/Common/Lovelace.hs#L176, but it is for _Byron_.

So I _guess_ you could setup a single-era cluster post-Byron with a different max supply. I presume this is what would calculate it: https://github.com/input-output-hk/cardano-ledger/blob/4347e54b2827090e898702ccfb93cc731b4a1909/eras/shelley/impl/src/Cardano/Ledger/Shelley/Genesis.hs#L434

For the proper `Cardano` mode (all eras), I suspect 45B still holds, though. So this might not be the most important thing to care about even if true.

HeinrichApfelmus replied:
I agree that we may want to avoid hard-coding the minimum and maximum value for `Coin` — they are actually configured in the Shelley genesis. (In fact, our local cluster has a higher maximum Ada quantity than 45B.)

In fact, I believe that we can remove `txOutMaxCoin` entirely — checking the maximum bound will never catch any user mistakes that will not manifest in a different way (e.g. inability to find inputs with sufficient funds).

When discussing this, Jonathan suggested to make a second PR in order to keep the scope of this PR to a refactoring (as opposed to a change in functionality).

jonathanknowles replied:
@Anviking I completely agree with your observation. This value could be different in custom testnets, so hard-coding it within the wallet primitive layer is of questionable value.

Though as @HeinrichApfelmus mentioned, it was my aim to avoid changing any behaviour in this PR.

I think we could/should consider making a ticket to capture this issue. I'll try to do that today.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -531,7 +531,7 @@ prefilterBlock b u0 = runState $ do
                 (Nothing, Outgoing) -> -- Byron
                     let
-                        totalOut = sumCoins (txOutCoin <$> outputs tx)
+                        totalOut = F.fold (txOutCoin <$> outputs tx)
```
I appreciate the clarity of the name `sumCoins` over `F.fold`, but I won't per-se object with this change either.

jonathanknowles replied:
> I appreciate the clarity of the name sumCoins over F.fold, but I won't per-se object with this change either.

I think my (not very strongly felt) objection to this stems from the fact that `sumCoins` seems specialized, but in reality it folds over the coins in a very similar way to how `F.fold` adds with the `Semigroup` and `Monoid` instances, so it _seems_ a bit superfluous.

But my objection is not a strong one! I would prefer the name `sum` though, if we were to retain it, since the current API of `Coin` is intended to be imported in qualified fashion.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs:
```diff
@@ -235,6 +259,67 @@ shrinkTxOut (TxOut a b) = uncurry TxOut <$> shrinkInterleaved
+--
+genTxOutCoin :: Gen Coin
+genTxOutCoin = frequency
+    [ (1, pure txOutMinCoin)
+    , (1, pure txOutMaxCoin)
+    , (8, Coin.fromNatural <$> chooseNatural
+        ( Coin.toNatural txOutMinCoin + 1
+        , Coin.toNatural txOutMaxCoin - 1
+        )
+      )
```
Would it be an idea to do something like this instead?
```haskell
n <- fromIntegral . getNonNegative <$> arbitrary @Int :: Natural
frequency
  [ (1, Coin.fromNatural $ txOutMinCoin + n)
  , (1, Coin.fromNatural $ txOutMaxCoin - n)
  , (8, Coin.fromNatural <$> chooseNatural (min, max))
  ] 
```
Motivation:
- More coverage _near_ boundaries
- Manually excluding the boundary cases from the full-range `chooseNatural` seems pointless

jonathanknowles replied:
I think we could definitely consider that!

Though perhaps in a separate PR?

(My aim with this PR was to keep it as a pure refactoring, as much as possible.)

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/CompatibilitySpec.hs:
```diff
@@ -409,7 +411,7 @@ prop_assessTokenBundleSize_enlarge b1' b2' =
         [ assess (b1 `TokenBundle.add` b2)
             === TokenBundleSizeExceedsLimit
-        , assess (b1 `TokenBundle.setCoin` maxBound)
+        , assess (b1 `TokenBundle.setCoin` txOutMaxCoin)
```
👌

Anviking reviewed (approved):
LGTM, I think the direction makes sense. (And thanks for tackling this!) I left a few comments from going through the commits start to finish, once.

HeinrichApfelmus commented on lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs:
```diff
@@ -69,26 +76,21 @@ import Quiet
     deriving stock (Ord, Eq, Generic)
     deriving (Read, Show) via (Quiet Coin)
 
 instance Semigroup Coin where
```
One of the most important things about the `Coin` abstraction is that it has a `Semigroup` and a `Monoid` instance which correspond to *summation*. Could you add corresponding comments in the Haddock documentation?

jonathanknowles replied:
Good point!

I've added some comments in 535d0dd85454b3d2ea1206c04ef13169c801c826.

HeinrichApfelmus commented on lib/shelley/cardano-wallet.cabal:
```diff
@@ -61,6 +61,7 @@ library
     , filepath
     , fmt
     , generic-lens
+    , int-cast
```
We use a bewildering variety of packages. At some point, we may want to document and consolidate our choices, so that writing mundane code becomes more streamlined (e.g. Text vs ByteString vs lazy ByteString; converting to/from hex). 😅 But I do not believe that the variety really hurts readability, so the status quo is fine as well.

jonathanknowles replied:
I do agree we use a very large variety of packages!

Though to some extent, perhaps this reflects the fragmentation and/or modularity within the Haskell package ecosystem?

In this case, I believe the `int-cast` package is rather well-designed and very useful for catching errors. We actually already do use it within `cardano-wallet`: this just adds it to some of the other build targets.

jonathanknowles replied:
> ByteString vs lazy ByteString

To some extent, I think our hand is forced here by the packages we rely on.

Historically, Aeson has used lazy `ByteString` by default. Though I've noticed that strict variants have recently been added for the most common functions.

jonathanknowles committed:
Use `Coin.toNatural` in `Api.Server.mkApiFee`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3034#discussion_r755176107

jonathanknowles committed:
Use `Coin.unsafeToWord64` in `computeUtxoStatistics`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3034#discussion_r755159414

jonathanknowles committed:
Add commentary relating to `Semigroup` and `Monoid` instances for `Coin`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3034#discussion_r755463847

jonathanknowles committed:
Fix integration test failures.

It turns out that on our integration testnet, the minimum quantity of lovelace
allowed in a transaction output is 0.

Hence we lower the `txOutMinCoin` constant to `Coin 0`.

In response to:

https://github.com/input-output-hk/cardano-wallet/pull/3034#issuecomment-977491169

jonathanknowles committed:
Fix `prop_assessTokenBundleSize_*` properties.

Due to changes in the definitions of generators, it's become necessary
to call `QC.resize` with a higher value in order to generate a range of
token bundle sizes that straddles the boundary between being "too small"
and "too large".
